[ENH] AptaTrans: training code and training schema

NennoMP opened the pull request:
Adds training code and training schema for AptaTrans, including code related to data loading and pre-processing. Specifically, training code via `lightning` for AptaTrans' deep neural network has already been implemente din #130. This PR will still involve implementing the same for pretraining AptaTrans' autoe-encoders.

This pull requests resolves #49 and is stacked on #63 (which should be merged before this).

## Data and model weights
Particularly big files (datasets, model weights) are hosted on [GC.OS Hugging Face](https://huggingface.co/gcos/).

satvshr commented on pyaptamer/utils/__init__.py:
```diff
@@ -1,9 +1,14 @@
 """Utils for the pyaptamer package."""
```
Just observed this, can you move everything out of utils into their respective files? It breaks the purpose of having util files for separate purposes if we move some functions to root import imo

satvshr replied:
This is something we had discussed in the weekly regarding the structure of utils, the main essence of the discussion was that init should only contain imports which are going to be used by users and algorithms for transformations, etc. Not private functions that serves the use case of a single algorithm, in that sense we should also move literally every function in utils to init, hence resulting in no structure for the file.

satvshr commented on pyaptamer/utils/augment.py:
```diff
@@ -0,0 +1,26 @@
+__author__ = ["nennomp"]
```
Schedule this for Tuesday then :D

satvshr commented on pyaptamer/base/_base_solver.py:
```diff
@@ -0,0 +1,309 @@
+from tqdm import tqdm
+
+
+class BaseSolver(ABC):
```
https://github.com/gc-os-ai/pyaptamer/issues/112

satvshr commented on pyaptamer/utils/_struct_to_aaseq.py:
> could you specify what the functions currently in the utils/__init__.py file are for and are they general purpose?

I meant in this thread (if you mistook it for docstring) sorry for not being clear!

satvshr commented on pyaptamer/datasets/data/dummy_data.csv:
```diff
@@ -0,0 +1,2 @@
+aptamer,protein,label
```
What do you think about structuring the `data` directory and moving files into their respective paths? SO `pdb` goes into `data/pdb` and so on. @fkiraly thoughts?
ps: If you agree and do this, you will have to change the paths of the helper functions too that I added in utils for file conversions.

NennoMP commented:
As mentioned above, we should discuss during a weekly which files to import to `utils.__init__.py` and which to keep in their dedicated submodules. I also think we should look at both AptaTrans and AptaNet utils and see if some methods overlap.

Info about what each util file does:
1. `utils._aptatrans_utils.py` contains utilities specific to AptaTrans, such as converting RNA and proteins sequences to a vector representation for the encoders' embedding;
2. `utils._augment.py` contains method for data augmentation. Currently, there is just one method that allows to add reverse complements of each sequence;
3. `utils._base.py` contains various generic methods. For instance, to generated all possible triplets combination for a given vocabulary or converting a DNA sequence to RNA.

Utilities in (2.) and (3.) are generic enough so that they could be used outside AptaTrans. Not sure if ver

satvshr commented:
> As mentioned above, we should discuss during a weekly which files to import to `utils.__init__.py` and which to keep in their dedicated submodules.

Agreed

> Info about what each util file does:

I think right now we can keep it private and when it becomes more generic, we make it public (this is what Franz had asked me to do too)

> I also think we should look at both AptaTrans and AptaNet utils and see if some methods overlap.

I will have a look and let you know

> Not sure if ver

I think you didnt complete your statement?

satvshr commented:
One more thing, you can integrate `pre-commit` to run automatically whenever you want to make a commit, unless you do not want that.

NennoMP commented:
> One more thing, you can integrate `pre-commit` to run automatically whenever you want to make a commit, unless you do not want that.

I will look into it, `pre-commit` is annoying me a lot recently :D. It says everything is ok, then I push and if I re-run it triggers errors idk...

satvshr commented:
> I will look into it, `pre-commit` is annoying me a lot recently :D. It says everything is ok, then I push and if I re-run it triggers errors that weren't spot before idk why.

Understandable 😄 look into running pre-commit locally, I think once you pip install it you have to run `pre-commit install` but I am not sure, if you need help with it let me know!

satvshr commented:
Changes I requested seem not to have been met, and the PR this was stacked on seems to have already been merged.
1. `model.py` and `pipeline.py` are not private.
2. `_rna.py` is still being mentioned in `__init__.py`

fkiraly commented on pyaptamer/utils/__init__.py:
```diff
@@ -1,19 +1,10 @@
 
+__author__ = ["nennomp"]
 __all__ = [
-    "dna2rna",
```
what is going on here? I think we should leave the exports as they are.

satvshr replied:
replied above.

fkiraly replied:
where?

fkiraly reviewed (changes requested):
The neural network code looks great, I am confused about the changes in the `utils` module. I think we should not be reorganizing existing code in the module in the same PR where we are doing something else.

Can you revert changes that do not concern AptaTrans, and if you want to reorganize `utils`, do it in another PR? We should probably also discuss whether these changes are necessary.

NennoMP commented on pyaptamer/mcts/_algorithm.py:
```diff
@@ -29,7 +29,7 @@ class MCTS(BaseObject):
     n_iterations : int, optional
         Number of iterations per round for the MCTS algorithm.
-    experiment : BaseExperiment, optional, default=None
+    experiment : Aptamer, optional, default=None
```
In this PR the experiment is still called `Aptamer`, to not confuse with how the experiments have been refactored in #121 to make `AptaNet `compatible with `MCTS`.

When #121 is merged I will need to address the conflicts in this PR I imagine.

satvshr commented on pyaptamer/aptatrans/_pipeline.py:
```diff
@@ -118,18 +111,15 @@ def _init_words(
 
     def _init_aptamer_experiment(self, target: str) -> Aptamer:
         """Initialize the aptamer experiment."""
-        # initialize the aptamer recommendation experiment
```
I think we had discussed making the comment the method docstring instead as it provided more clarity?

NennoMP replied:
> I think we had discussed making the comment the method docstring instead as it provided more clarity?

Yep, then the inline comment became redundant and I removed it.

fkiraly replied:
this is not an important line but I was still wondering why it gets removed. I do not mind if it gets removed, but it is just strange that it does.

satvshr commented:
> Can you revert changes that do not concern AptaTrans, and if you want to reorganize utils, do it in another PR? We should probably also discuss whether these changes are necessary.

I agree that the scope of this PR has massively increased and such concerns should be moved to a new PR, this (utils structure) was already discussed though in a weekly, as mentioned above.

I would also recommend making the notebook a separate PR dependent on this PR and making it as a priority, as it would help out with benchmarking `AptaTrans`.

fkiraly commented on pyaptamer/aptatrans/_pipeline.py:
```diff
@@ -6,24 +6,17 @@
 
 
 class AptaTransPipeline:
-    """AptaTrans pipeline for aptamer affinity prediction, by Shin et al.
```
I think references do not display properly in the header - may I suggest to revert?

fkiraly commented:
@NennoMP, would you mind describing how the content in this PR has been rearranged across PR (or whether you have decided to keep it in one place)

satvshr commented on pyaptamer/datasets/_loaders/_hf_loader.py:
```diff
@@ -0,0 +1,54 @@
+    print(f"Downloading {name}...")
+    try:
+        dataset = load_dataset(f"gcos/pyaptamer-{name}")
+        dataset = dataset["train"].to_pandas()
```
This will not work for file formats which are not dataframes or CSVs, was trying to make a loader for a fasta file and was going to use this until I noticed this.

satvshr commented on pyaptamer/datasets/_loaders/_hf_loader.py:
```diff
@@ -0,0 +1,54 @@
+
+    print(f"Downloading {name}...")
+    try:
+        dataset = load_dataset(f"gcos/pyaptamer-{name}")
```
This will not work for dataformats not noticed by huggingface, was trying to make a loader for a fasta file and was going to use this until I noticed this.

NennoMP commented:
> @NennoMP, would you mind describing how the content in this PR has been rearranged across PR (or whether you have decided to keep it in one place)

So, the content of the PR has remained pretty much the same except for training AptaTrans' deep neural network. We implemented that using `lightning` in #130. Data-preprocessing and loading, and overall training schema are still implemented in this PR. Additionally, I need to implement a `lightning` wrapper for pre-training AptaTran's auto-encoders, similarly to what we did in #130.

I updated the PR description to reflect this.

fkiraly reviewed (changes requested):
Reminder of change requests above.

* docstrings get changed, shortened or removed. Please revert unless there is a clear reason.
* please revert renaming of modules in `utils`

New requests/questions:

* `AptaTransPipeline` has a paramter `prot_words` which does not seem properly described, and its description is discrepant with the example. Further, should there not be a default for this?
* I think there should be a `predict` method as well. Is it just a matter of renaming `predict_api`?

fkiraly commented:
can you kindly resolve the merge conflicts in the stack? Please update from `main`, the PR #63 has been merged.

NennoMP commented:
> Reminder of change requests above.
> 
> * docstrings get changed, shortened or removed. Please revert unless there is a clear reason.
> * please revert renaming of modules in `utils`
> 
> New requests/questions:
> 
> * `AptaTransPipeline` has a paramter `prot_words` which does not seem properly described, and its description is discrepant with the example. Further, should there not be a default for this?
> * I think there should be a `predict` method as well. Is it just a matter of renaming `predict_api`?

Still need to address this. Will do and then update this comment.
